nixos/modules/programs/ssh.nix

```diff
@@ -26,10 +26,103 @@ let
     ++ builtins.map pkgs.copyPathToStore cfg.knownHostsFiles;
 
+
+  settingsNameValueType = with lib.types; submodule {
+    freeformType = lazyAttrsOf (nullOr (oneOf [ singleLineStr int bool (listOf singleLineStr) ]));
+
+    options = {
+      priority = lib.mkOption {
+        description = ''
+          Order of this Host/Match block in relation to the others in the same type.
+          The semantics are the same as with `lib.mkOrder`. Smaller values have
+          a greater priority.
+        '';
+        type = lib.types.int;
+        default = 1000;
+      };
+
+      ForwardX11 = lib.mkOption {
+        type = lib.types.nullOr lib.types.bool;
+        default = false;
+        description = ''
+          Whether to request X11 forwarding on outgoing connections by default.
+          If set to null, the option is not set at all.
+          This is useful for running graphical programs on the remote machine and have them display to your local X11 server.
+          Historically, this value has depended on the value used by the local sshd daemon, but there really isn't a relation between the two.
+          Note: there are some security risks to forwarding an X11 connection.
+          NixOS's X server is built with the SECURITY extension, which prevents some obvious attacks.
+          To enable or disable forwarding on a per-connection basis, see the -X and -x options to ssh.
+          The -Y option to ssh enables trusted forwarding, which bypasses the SECURITY extension.
+        '';
+      };
+    };
+  };
+
+  commaSeparatedSettings = [
+    "CASignatureAlgorithms"
+    "Ciphers"
+    "HostKeyAlgorithms"
+    "HostbasedAcceptedAlgorithms"
+    "KexAlgorithms"
+    "MACs"
+    "PreferredAuthentications"
+    "ProxyJump"
+    "PubkeyAcceptedKeyTypes"
+  ];
+  spaceSeparatedSettings = [
+    "DynamicForward"
+    "GlobalKnownHostsFile"
+    "IPQoS"
+    "LocalForward"
+    "RemoteForward"
+    "UserKnownHostsFile"
+    "CanonicalDomains"
+    "CanonicalizePermittedCNAMEs"
+    "SendEnv"
+    "SetEnv"
+    "IdentityFile"
+    "Include"
+    "LogVerbose"
+    "PermitRemoteOpen"
+    "RevokedHostKeys"
+  ];
+
+  # Renders a single value for openssh `ssh_config` with key and value and indented by 2 spaces.
+  renderSingleValue = settingName: settingValue: "  ${settingName} " +
+    (if lib.isInt settingValue then toString settingValue
+    else if lib.isString settingValue then settingValue
+    else if lib.isBool settingValue && settingValue then "yes"
+    else if lib.isBool settingValue && !settingValue then "no"
+    else if lib.isList settingValue && builtins.elem settingName commaSeparatedSettings then builtins.concatStringsSep "," settingValue
+    else if lib.isList settingValue && builtins.elem settingName spaceSeparatedSettings then builtins.concatStringsSep " " settingValue
+    else if lib.isList settingValue then throw "list value for unknown key ${settingName}: ${(lib.generators.toPretty {}) settingValue}"
```

Comment:
I do wonder if it isn't a little nicer to do the `isX` invocations once:

```nix
if isBool settingValue then
  if settingValue then "yes" else "no"
else if isList settingValue then
  if elem settingName commaSeparatedSettings then "..."
  else "..."
else "..."
```

Since this isn't a hot code-path I'm not even sure if that makes such a big different, so this is optional.